### ISSUE TITLE
Experiment: Discord Game Invites

### ISSUE THREAD

**Issue by jacobtread** [PocketRelay/ServerRust#53]
## Description

See if its possible to use the [Discord game sdk](https://discord.com/developers/docs/game-sdk/sdk-starter-guide) to implement an invite system independently of the origin/ea overlay, this would improve the invite system reliability as the current reliability of the existing overlay is flakey and its unknown how long it will continue to work.

If this experiment is a success #32 can be closed

## Checklist
- [ ] Game invites are integrated with discord
- [ ] Players can join each other through discord

**Comment by SoraBytes** [comment]
I don't use discord nor my close friends for communications or games. Relying on it for an external invite system would eliminate people like us from being able to use an in-game invite system. I would rather suggest implementing a different approach or providing two methods if you're set on implementing discord.

I would suggest something like you add to the relay client a "connect" box that contains another user's in-game name and a "password" box that -must- match on the other end and validated by the server. 

Example:
Host: sets password -> forwards it to server
Server: stores password for host using its associated accout name
Client: enters name/password -> forward to server
Server: validates name/password (accept|decline)
Accept: Host is forwarded client joining packet and Client is forwarded game info
Decline: send Client wrong info (e.g. cause a prompt about not able to join or something, or just flatout ignore it)

**Comment by jacobtread** [comment]
Hey, I'm not too sure if im going down the discord route yet as I'm not sure if I can make the discord add work with this. This is an experiment so its not set in stone and things are bound too change. This experiment was added as I've seen a large portion of the user's and ME3 playerbase reside on discord communities

At the moment I'm working on modifying the game in order to make it possible to obtain the player details and call the nessicary client(me3) game joining functions. Because of the way the client(pr) is structured currently it has no idea about authentication or anything that's happening on the game client as its only a proxy, attempting to modify this to intercept authentication from the server would add a great performance penalty.

Because of this invite changes won't be available to the original external client tool   instead im only planning on supporting them for the client plugin, as with the client plugin im able to directly access the game(me3) objects and interact, which simply isn't possible without being injected into the game (the normal tool isn't injected like the plugin and its impossible to force a client to join from the server side so the plugin version must be used)

I may find other solutions while I'm working on this but ill definitely keep your suggested method in mind.

**Comment by LocalBoomer** [comment]
Call me the necromancer necroing this issue, but...

1. Discord game SDK is dead. Discord is moving to their [Social SDK](https://discord.com/developers/docs/discord-social-sdk/overview). Probably won't make a difference but still.
2. If possible, I think the best solution would be a 'plugin extendable' solution of using the PR client to manage invites. Being able to just use the program itself to invite people but allowing integrations with things like Discord to be added on top so people can still invite regardless of what social platform they want to use or not use. (Maybe using a link re-direct solution so people can just send invite links to each other in any form that open with the client?

**Comment by jacobtread** [comment]
Yeah, thats probabbly the way to go. This issue hasn't made much progress yet. I haven't worked on this in a while and I don't believe I got too far with the implementation for this (Atleast theres no discord side code, I believe I was only working on the core game side part of this so far) I don't think I got anywhere in terms of an actual implementation but I'll keep in mind your idea of redirects, 
deep-linking something like a pocket-relay://join-game?id={GAME}&key={SECRET} could probabbly be something thats possible I'm pretty sure windows lets us register deep links like that at runtime still more things to sort out like how invites would be created in game etc. But i'll keep that all in mind if I revisit this
